python_modules/libraries/dagster-dbt/dagster_dbt/asset_defs.py

```diff
@@ -222,26 +232,47 @@ def _events_for_structured_json_line(
 
 def _stream_event_iterator(
     context: OpExecutionContext,
-    dbt_resource: DbtCliResource,
+    dbt_resource: Union[DbtCli, DbtCliResource],
     use_build_command: bool,
     node_info_to_asset_key: Callable[[Mapping[str, Any]], AssetKey],
     runtime_metadata_fn: Optional[
         Callable[[OpExecutionContext, Mapping[str, Any]], Mapping[str, RawMetadataValue]]
     ],
     kwargs: Dict[str, Any],
+    manifest_json: Mapping[str, Any],
 ) -> Iterator[Union[AssetObservation, Output]]:
     """Yields events for a dbt cli invocation. Emits outputs as soon as the relevant dbt logs are
     emitted.
     """
-    manifest_json = check.not_none(dbt_resource.get_manifest_json())
-
-    for parsed_json_line in dbt_resource.cli_stream_json(
-        command="build" if use_build_command else "run",
-        **kwargs,
-    ):
-        yield from _events_for_structured_json_line(
-            parsed_json_line, context, node_info_to_asset_key, runtime_metadata_fn, manifest_json
+    if isinstance(dbt_resource, DbtCliResource):
+        for parsed_json_line in dbt_resource.cli_stream_json(
+            command="build" if use_build_command else "run",
+            **kwargs,
+        ):
+            yield from _events_for_structured_json_line(
+                parsed_json_line,
+                context,
+                node_info_to_asset_key,
+                runtime_metadata_fn,
+                manifest_json,
+            )
+    else:
+        if runtime_metadata_fn is not None:
+            raise DagsterDbtError(
+                "The runtime_metadata_fn argument on the load_assets_from_dbt_... functions is not"
+                " supported when using the DbtCli resource. Use the @dbt_assets decorator instead"
+                " of load_assets_from_dbt_... if you want control over what metadata is yielded at"
+                " runtime."
+            )
```

Comment:
> Can we raise it at definition time?

I'm not sure where we'd do this. I agree that it's possible to know this before executing the run. However, at the time `load_assets_from_dbt...` is invoked, we don't know what resources will be provided to it.

Reply:
> We could support runtime_metadata_fn if we want to using DbtCli.

Yeah I figured it would just be a waste if we're planning to deprecate it anyway.